Devops restructure python

sbc64 opened the pull request:
This PR contains the ground work for how the devops container management, integration tests, and testnets will look like.

There are five new python scripts:
- config.py
- build_containers.py
- testnet.py
- integration_test.py
- push_containers.py

### config.py:

This script is what determines the variables that the other scripts will use. The center piece is the `CONFIG` dictionary which is the source of all the environment variables that can be used to tweak the variables of the scripts.

Steps: 
1. check if any environment variables are set.
2. set all of the git hashes and branches that are missing.
3. determine whether warped is needed.

For a list of all the environment variables available look at `ops/README.md`

### build_containers.py

Checks with docker hub to see if the containers exist and have the same sha256 digest. If they don't then it builds the container with either the warped epochs or not. Warped epoch is determined in `config.py`.

Warped epoch is turned on by default on all non master branches. it can be turned off at any moment with `WARPED=false` environment variable.

The containers built are:
- `hashcloak/authority`
- `hashcloak/server`
- `hashcloak/meson`

The steps are:
1. Determine whether to pull or build the containers. If pull then the following steps are skipped.
2. Creates a temporary dockerfile that sets warped, if needed.
3. Builds the containers.

### testnet.py

A script that concatenates a docker-compose.yml file manually (I did not want to add an external python dependency). It uses the variables set in `config.py`. It makes a directory in tempdir + `meson-testnet`.

The steps are:
1. Get ip address.
2. Create temporary testnet directory and run genconfig
3. Obtain the public key of atuhority, and the port numbers neeeded for the dokcerfile.
4. Concatenates a `docker-compose.yml` file
 
### integration_test.py

Runs the integration tests that exist in `github.com/hashcloak/Meson-client/integration`.

Currently only runs the ethereum integration test.

### push_containers.py

Pushes the docker containers.

### .travis.yml
Travis just runs `make integration_test` and if successful pushes the new containers to docker hub.

Commit:
Saving combined key names

Commit:
Saving unique key path config

Commit:
Working dictionary expand

Commit:
Update local git functions

Commit:
Working get environment variables

Commit:
Working setNestedValue

Commit:
Values that need to be fetch working

Commit:
Add warped build check

Commit:
Add container api requests

Commit:
Add comparison of remote containers

Commit:
Update Makefile to use python scripts

Commit:
Update get upstream

Commit:
Working get_upstream.py

Commit:
Update encoding call

Commit:
Improved logs and silence git checkout

Commit:
Update Dockerfile

Commit:
Update .gitignore

Commit:
Removed comment

Commit:
Add build meson python

Commit:
Update go modules

Commit:
Add tesnet script

Commit:
Update comment from makef

Commit:
Cleaned up for better readability

Commit:
Update Makefile and travis. Init for integration test

Commit:
Removed encoding parameter from checkout_output

This parameter is not availabel for check_output, but it
is available for subprocess.run()

Restructed DEFAULT_VALUEs to CONFIG

Add push containers

Removed utf-8 since it is default for .decode()

Improved get_upstream and fix the named tag pull

Modularized imports for integration_test.py

Changed all check_output() to run()

Run has a live feed on the stdout while
check_output doesn't print until the command
is finished. check_output made the script feel
like it was frozen

Combined repo checkout logic into a function

Commit:
Removed code that wasn't neccesary and made it simpler

Commit:
Add README.md

Commit:
Fixed typos

Commit:
Removed extra spaces

Mikerah commented:
First things first, before I review, can you update your PR description to describe what the expected behaviour is and what steps I should do to ensure that I get it right? It'll make for a more cohesive PR review and should minimize the back and forth.

Commit:
Improved readability of config.py

Commit:
Removed comment

Commit:
Update links

Commit:
Improved readbility of ops docs

Commit:
Reduced complexity of compareRemoteContainers

Mikerah commented on README.md:
````diff
@@ -10,11 +12,103 @@ A Library for adding cryptocurrencies to Meson.
+To start a testnet run the following command:
+
+```
+make tesnet
````
When I run this, I get an error that indicates that I need to run one of `docker swarm init` or `docker swarm join`

Mikerah commented on ops/README.md:
```diff
@@ -0,0 +1,124 @@
+# How to use the ops directory:
```
Does this actually have to be in this README? Why not just add it to the overall docs for the project in the Meson repo?
READMEs after all should be short and to the point.

sbc64 replied:
I made a new issue to track this: https://github.com/hashcloak/Meson/issues/100

Commit:
Add docker swarm init

Mikerah commented on ops/build_containers.py:
```diff
@@ -0,0 +1,96 @@
+
+def getContainerInfo(name, tag):
+    if not doesContainerExistsInCloud(name, tag):
+        return {}
```
I'm debating whether returning an empty dictionary here is a good approach given that this function is being used elsewhere to in `if` statements to determine if the container exists. Either return a tuple i.e. `(False, {})` to indicate whether the value exists and the actual value or return a meaningful exception so that other code that uses this function can catch it.

sbc64 replied:
empty dicts are false. I don't see a point of adding a tuple when we could clarify in the docstrings what this function does.

The user of this fucntion expects to get some information, so either returning an empty dict or raising an error that the caller can handle are the best way to inform the caller that there was no information on the requested container.

Mikerah replied:
I am aware that empty objects are false-y but that doesn't mean it's necessarily good practice to do so. Not to mention the extra overhead in generating the empty object for this purpose.

So, either raise an exception or the tuple trick I wrote in the previous comment.

sebohe replied:
Overhead for python? or overhead for the developer handling the function?

Mikerah commented on ops/build_containers.py:
```diff
@@ -0,0 +1,96 @@
+    return loads(urllib.request.urlopen(url).read().decode())
+
+def compareRemoteContainers(ctr1, ctr2):
+    info1 = getContainerInfo(ctr1.split(":")[0], ctr1.split(":")[1])
```
I can tell that you are trying to retrieve the tag and names for two different, potentially existing remote containers but this is hard to read. Use properly named variables.

Mikerah commented on ops/build_containers.py:
```diff
@@ -0,0 +1,96 @@
+def compareRemoteContainers(ctr1, ctr2):
+    info1 = getContainerInfo(ctr1.split(":")[0], ctr1.split(":")[1])
+    info2 = getContainerInfo(ctr2.split(":")[0], ctr2.split(":")[1])
+    if not info1 or not info2:
```
See comment that I left for `getContainerInfo` above.

Mikerah commented on ops/build_containers.py:
```diff
@@ -0,0 +1,96 @@
+    return info1['images'][0]['digest'] == info2['images'][0]['digest']
+
+def buildContainer(container, tag, dockerFile, path):
+    print("\nLOG: Building {}:{}\n".format(container, tag))
```
As this print statement is solely for logging, have a default parameter to disable logging by default and if needed, one can enable logging by explicitly, setting the argument to `True`

Mikerah commented on ops/build_containers.py:
```diff
@@ -0,0 +1,96 @@
+    ], check=True)
+
+def reTag(container, tag1, tag2):
+    print("\nLOG: Tagging {} {} -> {}\n".format(container, tag1, tag2))
```
Same as in previous comment.

Mikerah commented on ops/build_containers.py:
```diff
@@ -0,0 +1,96 @@
+
+    return tmpdf, repoPath
+
+for key in ["AUTH", "SERVER", "MESON"]:
```
Instead of having an array, you can just do `CONFIG.keys()`.

sbc64 replied:
Not all keys in CONFIG are containers. But I can create a new dict in config that has the containers.

Mikerah replied:
Restructure `CONFIG` differently.

Mikerah commented on ops/testnet.py:
```diff
@@ -0,0 +1,153 @@
+
+generateClientTOML(testnetConfDir, ip, startingPortNumber, authorityPublicKey)
+
+composeYMLFile = """
```
Maybe create a utils file where you have all these cookie-cutter file generation utilities and other stuff across files.

Mikerah reviewed (changes requested):
Did an initial pass. A few things:
- Use type annotations. It makes it easier to know what the types of the parameters are in functions
- Add doc strings. Although the names of all the functions make sense, it much easier to read docstrings first and determine what it takes and what is returned.

Will do a subsequent PR when these changes have been resolved.

Commit:
Restructured CONFIG and use docker build args

Commit:
Add type annotation and docstrings

Change the build arg value to use config instead of default "warped"

Add utils

Remove check_output, it is a python 3.7 feature

Add a special log

Add sorted to ensure build order.

Fix travis python3.5 missing keyword capture_output

Add BUILD environment var to force build

Add type annotations and add error handling for comapareRemoteContainers

Removed generalized error for a more specific one

Reduced scope of error to catch a specific case and not catch all

Implemented some self review improvements

Cleaned up a  bit more

sbc64 commented:
Some of the major changes since the last review:

- The Dockerfile string replacing and string juggling has been removed since warped build is now supported by docker build args.
- A few more environment variables that can help with the build process:
   - `BUILD`, turn on to force the build of the containers
  - `TEST_ATTEMPTS`, number of attempts to run the integration tests

Review implementations:
- Added option to toggle logs on or off with `LOG` environment variable. Default is off.
- Instead of returning empty dict `getContainerInfo` now raises an error.
- The README in the ops directory will get moved to docs in `Meson`
- Docker swarm gets initialized if it is not present.
- Docstrings
- Type annotations
- Combined the yaml string concatenation/formatting into a single function

Commit:
Combined yml generation into one generic function

Commit:
Add missing quote and improve readability

Commit:
Fix typo

Commit:
Now parsing the paths of the genconfig tool

Commit:
Combined docker compose yaml concatenating logic

Commit:
Removed port counting and juggling

Commit:
Add docstrings

Commit:
Reduced hardcoded strings

Commit:
Return the full path of the toml instead of just the conf dir

Mikerah commented on ops/build_containers.py:
```diff
@@ -0,0 +1,96 @@
+            buildContainer(repo["CONTAINER"], repo["HASHTAG"], dockerFile, repoPath)
+            reTag(repo["CONTAINER"], repo["HASHTAG"], repo["NAMEDTAG"])
+
+main()
```
Add `if __name__ == '__main__'` here

Mikerah commented on ops/build_containers.py:
```diff
@@ -0,0 +1,96 @@
+
+def buildContainer(container: str, tag: str, dockerFile: str, path: str) -> None:
+    log("Building {}:{}".format(container, tag))
+    args = [
```
There should be a check for whether docker is installed on the system. If it's not return an error.

Mikerah commented on ops/build_containers.py:
```diff
@@ -0,0 +1,96 @@
+            "server={}".format(CONFIG["REPOS"]["SERVER"]["NAMEDTAG"])
+        ])
+    log(args)
+    run(args, check=True)
```
`try/catch` `CalledProcessError` for graceful failure.

sbc64 replied:
Graceful failure is not always something we want. I need to think better but failures make  sure the tests don't pass and we can take a look at what is happening.

Not saying that I'm against your proposal, I just need to look at it more and see where it is necessary.

Mikerah replied:
Maybe graceful failure isn't the term here but we also need to consider the fact that devops people will be using these scripts as well and it would make their lives easier if we have descriptive messages telling them what might be going wrong. We can't only consider integration tests.

Mikerah commented on ops/build_containers.py:
```diff
@@ -0,0 +1,96 @@
+
+def reTag(container: str, tag1: str, tag2: str) -> None:
+    log("Tagging container {} with {} -> {}".format(container, tag1, tag2))
+    run([
```
Catch `CalledProcessError` here as well.

Mikerah commented on ops/config.py:
```diff
@@ -0,0 +1,122 @@
+    "BUILD": "",
+}
+
+gitHashLength=7
```
This is a constant. So, use the convention of all caps and near the beginning of the file.

Mikerah commented on ops/config.py:
```diff
@@ -0,0 +1,122 @@
+    """
+    tempList = []
+    for key, value in dictionary.items():
+        if type(value) == dict:
```
I wonder if the following simplification is actually a simplification:
Instead of 
```{python}
for item in expandDictionary(value):
    tempList.append(key+separator+item)
```
you can do,
```{python}
tempList.extend([key+separator+item for item in expandDictionary(value)])
```

sbc64 replied:
Yeah, I think i did that just that the line was a bit too long for my taste. Willa dd it now.

Mikerah commented on ops/config.py:
```diff
@@ -0,0 +1,122 @@
+        repo["NAMEDTAG"] = "warped_"+repo["BRANCH"] if CONFIG["WARPED"] else repo["BRANCH"]
+        repo["HASHTAG"] = "warped_"+repo["GITHASH"] if CONFIG["WARPED"] else repo["GITHASH"]
+
+main()
```
Again check the namespace with `__name__`

Mikerah commented on ops/integration_test.py:
```diff
@@ -0,0 +1,44 @@
+    CONFIG["TEST"]["CLIENTCOMMIT"]
+)
+
+warpedBuildFlags='-ldflags "-X github.com/katzenpost/core/epochtime.WarpedEpoch=true -X github.com/katzenpost/server/internal/pki.WarpedEpoch=true"'
```
Is this still needed now that it can easily be toggle directly from the `Dockerfile` itself?

sbc64 replied:
Yes, it is needed for the integrations tests. If it is not present then the client requests for a different epoch.

Mikerah commented on ops/testnet.py:
```diff
@@ -0,0 +1,162 @@
+
+
+def getMixnetPort(path: str) -> str:
+    """Gets mixnet port number from a given katzenpost.toml file"""
```
You should be able to use a toml parser tool here instead of manually doing it yourself.

sbc64 replied:
Sure I can add the toml parser. But then pip it is one more additional dependency that needs to be installed and handled. I tried avoiding installing pip packages because then i would have to deal with it for just a toml or yaml parser.

I have no problem using pip, virtualenv, or nix-shells, they all work great. I just seem to have a hard time of making it reliable on other peoples system for just a string parser.

sbc64 replied:
By dealing with it i mean that I have to tell users how to install pip or tell them to use their prefer virtual env solution. The docker swarm system I'm using is already causing some friction from what I can see in your other comments and I would prefer to avoid another dependency manager for this PR

Mikerah commented on ops/testnet.py:
```diff
@@ -0,0 +1,162 @@
+                return line.split('"')[1].split(":")[1]
+
+def getUserRegistrationPort(path: str) -> str:
+    """Gets the user registration port from a given katzenpost.toml file"""
```
Same here

Mikerah commented on ops/testnet.py:
```diff
@@ -0,0 +1,162 @@
+    """Starts Docker stack deploy. Starts a docker swarm if there isn't one"""
+    output = check_output(["docker", "info"])
+    if "Swarm: inactive" in output.decode():
+        run(["docker", "swarm", "init", "--advertise-addr={}".format(ip)], check=True)
```
Catch `CalledProcessError`

Mikerah commented on ops/testnet.py:
```diff
@@ -0,0 +1,162 @@
+        "mixnet",
+        ], check=True)
+
+    run(["docker", "service",  "ls", "--format", '"{{.Name}} with tag {{.Image}}"'])
```
Same here

Mikerah reviewed (changes requested):
Overall, looks good. A few general comments:
- Convert camel case to snake case. It's the pythonic way of naming functions
- Based on the instructions in the `README`, I couldn't get the testnet running. I kept on getting a `CalledProcessError` when I tried the following make commands `make build_meson` and `make testnet`. I should add that I called `make clean` first as well. Maybe add more details around this.

Commit:
Add better list handling for expand dictionary

Commit:
Add namespace, snake case, and docker check

Commit:
Add docker check

Commit:
Add docker check and setup_config function

Commit:
Add docker error logs

Commit:
Add missing assigment

Commit:
Add namespace to push containers

Mikerah commented on ops/utils.py:
```diff
@@ -0,0 +1,64 @@
+def log(message: str, err=False) -> None:
+    """Logs a message to the console with purple.
+    If err is True then it will log with red"""
+    color = '\033[0;31m' if err else '\033[0;35m'
```
Why purple? Why not blue?

sbc64 replied:
Any color really.

Mikerah replied:
For maximal matrix effects, it should really be green.

Mikerah reviewed (changes requested):
![new_error_make_integration_tests](https://user-images.githubusercontent.com/4249799/83278205-350be500-a1a1-11ea-9ea4-0cd0cb8b0173.png)
I would have approved this PR but I got the above error when calling `make integration_tests`.

Commit:
Add check for private key

Commit:
Change log color to green

Commit:
Add ignore error conditional

Commit:
Improve error logic

Commit:
Fix function name

Mikerah commented:
The main thing to add to this PR to make it go through would be more instructions on running the integration tests.

Commit:
Update docs with instructions

sbc64 commented:
Resolved your latest request: https://github.com/hashcloak/Meson-plugin/pull/26/commits/7e628c6bab7fc93495e2ac9fadb586ad49cf2703

The environment variables list link will be updated with this issue/pr: https://github.com/hashcloak/Meson/issues/90

Commit:
Add addtional logs

Commit:
Add missing import
